sparkfe/design_of_fetool.md

```diff
@@ -0,0 +1,82 @@
+- Start Date: 2021-05-13
+- Target Major Version: 1.0
+- Reference Issues: https://github.com/4paradigm/SparkFE/issues/69
+- Implementation PR: 
+
+# Summary
+
+We should provide some tools with easy-to-use API like Python for users to meet the featue extraction requirements like creating FEDB DDL.
+
+
+# Basic example
+
+Users can use fetool command to create FEDB DDL of creating tables.
+
+```
+fetool gen_ddl sql.yaml
+```
+
+Users can use fetool command to finish the following tasks without developing.
+
+```
+fetool csv_to_parquet /csv_files /parquet_files
+fetool sample_parquet /parquet_files
+fetool inspect_parquet /parquet_files
+fetool check_skew /parquet_files
+fetool benchmark 'spark-submit --master local /pyspark_app.py'
+```
+
+# Motivation
+
+Now users can use SparkFE for feature extraction with SQL API. Development is required since it's the library of distributed computing and do not solve problems without specific SQLs. 
+
+However, there are some common tools which may be used for general feature extraction scenarios. For examples, we may sample the input dataset and check if its distribution is balanced. These tools are general and useful for feature extraction which may reduce the cost of development if we want to use SparkFE for AI. If we inspect the distribution of dataset in advance, the window skew optimization may use the distribution to achieve better performance.
+
+Therefore, providing the common tools for feature extraction is useful for developers. The easiest way to use is commad-line and we should provide Python and Java API for different developers.
+
+# Detailed design
+
+We want to use Python to wrap the feature extraction tools since Python is easy to use and has integrated with other programming languages.
+
+Users can install this tool with `pip` and all the functions can be called with command-line tool and Python functions. The fetool should be the standard Python package and command-line tool. There are some kinds of secnarios which fetool should support.
+
+* For data processing jobs like converting dataset and sampling dataset, we can use PySpark API which can submit the jobs in Python programming language.
+* For benchmark tool which requires running multiple jobs with different Spark distributions, we may use `subprocess` api to invoke the shell commands.
```

Comment:
doest SparkFE needed for installation as a dependency ?

Reply:
Some tools require to use SparkFE like `fetool benchmark` which need to download SparkFE distribution in advance.